File: sdks/python/relayhub/resources/notifications.py
from __future__ import annotations

from typing import Any

from ..http import RequestOptions, Transport
from ..types import AlertEventOut, AlertRuleOut, TestAlertResponse


class NotificationsResource:
    """
    "Notifications" maps to RelayHub's alert rules: Slack, Discord, webhook, or
    email notifications fired when an endpoint's failure rate crosses a threshold.
    There is no separate /notifications API in the backend -- this resource is
    kept under that name because that's the developer-facing concept it serves,
    while every method below hits the real /v1/alerts/* routes.
    """

    def __init__(self, transport: Transport) -> None:
        self._transport = transport

    def create_rule(
        self,
        *,
        condition_type: str,
        channel: str,
        channel_config: dict[str, Any],
        severity: str = "warning",
        threshold_config: dict[str, Any] | None = None,
        throttle_window_minutes: int | None = None,
        is_enabled: bool = True,
        options: RequestOptions | None = None,
    ) -> AlertRuleOut:
        """POST /v1/alerts/rules"""
        body = {
            "condition_type": condition_type,
            "severity": severity,
            "channel": channel,
            "channel_config": channel_config,
            "threshold_config": threshold_config or {},
            "throttle_window_minutes": throttle_window_minutes,
            "is_enabled": is_enabled,
        }
        return self._transport.request("POST", "/v1/alerts/rules", body, options)

    def list_rules(self, options: RequestOptions | None = None) -> list[AlertRuleOut]:
        """GET /v1/alerts/rules"""
        return self._transport.request("GET", "/v1/alerts/rules", None, options)

    def update_rule(self, rule_id: str, *, options: RequestOptions | None = None, **fields: Any) -> AlertRuleOut:
        """PATCH /v1/alerts/rules/{id}"""
        return self._transport.request("PATCH", f"/v1/alerts/rules/{rule_id}", fields, options)

    def delete_rule(self, rule_id: str, options: RequestOptions | None = None) -> None:
        """DELETE /v1/alerts/rules/{id} -- 204 No Content on success."""
        return self._transport.request("DELETE", f"/v1/alerts/rules/{rule_id}", None, options)

    def test_rule(self, rule_id: str, options: RequestOptions | None = None) -> TestAlertResponse:
        """POST /v1/alerts/rules/{id}/test -- fires a test notification through the rule's configured channel."""
        return self._transport.request("POST", f"/v1/alerts/rules/{rule_id}/test", None, options)

    def history(
        self, *, condition_type: str | None = None, limit: int = 50, offset: int = 0, options: RequestOptions | None = None
    ) -> list[AlertEventOut]:
        """GET /v1/alerts/history"""
        opts = options or RequestOptions()
        opts.query = {**(opts.query or {}), "condition_type": condition_type, "limit": limit, "offset": offset}
        return self._transport.request("GET", "/v1/alerts/history", None, opts)

File: sdks/node/src/resources/notifications.ts
import type { Transport, RequestOptions } from "../transport.js";
import type { AlertEventOut, AlertRuleOut, TestAlertResponse } from "../types.js";

/**
 * "Notifications" maps to RelayHub's alert rules: Slack, Discord, webhook, or
 * email notifications fired when an endpoint's failure rate crosses a threshold.
 * There is no separate `/notifications` API in the backend -- this resource is
 * kept under that name because that's the developer-facing concept it serves,
 * while every method below hits the real `/v1/alerts/*` routes.
 */
export class NotificationsResource {
  constructor(private readonly transport: Transport) {}

  /** POST /v1/alerts/rules */
  createRule(
    params: {
      condition_type: string;
      severity?: "info" | "warning" | "critical";
      channel: string;
      channel_config: Record<string, unknown>;
      threshold_config?: Record<string, unknown>;
      throttle_window_minutes?: number;
      is_enabled?: boolean;
    },
    options?: RequestOptions
  ) {
    return this.transport.request<AlertRuleOut>("POST", "/v1/alerts/rules", params, options);
  }

  /** GET /v1/alerts/rules */
  listRules(options?: RequestOptions) {
    return this.transport.request<AlertRuleOut[]>("GET", "/v1/alerts/rules", undefined, options);
  }

  /** PATCH /v1/alerts/rules/{id} */
  updateRule(
    ruleId: string,
    params: Partial<{
      severity: "info" | "warning" | "critical";
      channel: string;
      channel_config: Record<string, unknown>;
      threshold_config: Record<string, unknown>;
      throttle_window_minutes: number;
      is_enabled: boolean;
    }>,
    options?: RequestOptions
  ) {
    return this.transport.request<AlertRuleOut>("PATCH", `/v1/alerts/rules/${ruleId}`, params, options);
  }

  /** DELETE /v1/alerts/rules/{id} -- 204 No Content on success. */
  deleteRule(ruleId: string, options?: RequestOptions) {
    return this.transport.request<void>("DELETE", `/v1/alerts/rules/${ruleId}`, undefined, options);
  }

  /** POST /v1/alerts/rules/{id}/test -- fires a test notification through the rule's configured channel. */
  testRule(ruleId: string, options?: RequestOptions) {
    return this.transport.request<TestAlertResponse>("POST", `/v1/alerts/rules/${ruleId}/test`, undefined, options);
  }

  /** GET /v1/alerts/history */
  history(params?: { condition_type?: string; limit?: number; offset?: number }, options?: RequestOptions) {
    return this.transport.request<AlertEventOut[]>("GET", "/v1/alerts/history", undefined, { ...options, query: { ...params } });
  }
}
